### PR TITLE
Added an experimental convert scripts option for godot3 export

### PR DESCRIPTION
**Pull request by marcelofg55**
If enabled this option will convert gdscripts using regexp. So far it converts:

- _fixed_process( => _physics_process(
- _pos( => _position(
- KEY_RETURN => KEY_ENTER
- RawArray() => PoolByteArray()
- Vector2Array() => PoolVector2Array()
- .get_opacity() => .get_modulate().a
- .set_opacity(var) => .modulate.a = var
- .connect( => .connect_to_host( _// (A HTTPClient object is detected first)_
- ReferenceFrame => Control
- var.type == InputEvent.KEY => var.get_name() == "InputEventKey"
- var.type == InputEvent.MOUSE_MOTION => var.get_name() == "InputEventMouseMotion"
- var.type == InputEvent.MOUSE_BUTTON => var.get_name() == "InputEventMouseButton"
- var.type == InputEvent.JOYSTICK_MOTION => var.get_name() == "InputEventJoypadMotion"
- var.type == InputEvent.JOYSTICK_BUTTON => var.get_name() == "InputEventJoypadButton"
- is_move_and_slide_on_floor() => is_on_floor()
- is_move_and_slide_on_ceiling() => is_on_ceiling()
- is_move_and_slide_on_wall() => is_on_wall()

### CODE REVIEW

**Comment by Zylann** [comment]
set_opacity() conversion must use current color, not white. (If you assume that any set_opacity function is called on an object having a modulate)

**Comment by reduz** [comment]
Wow

On Oct 7, 2017 10:52 PM, "Marc" <notifications@github.com> wrote:

set_opacity() conversion must use current color, not white. (If you assume
that any set_opacity function is called on an object having a modulate)

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/godotengine/godot/pull/11863#issuecomment-334977264>,
or mute the thread

**Comment by Zylann** [comment]
You could also rename "len" identifiers into "length", because "len" is now a reserved function keyword (if there is no other variable in the same scope named "length" already)

**Comment by Zylann** [comment]
@reduz what's going on? :D

**Comment by Zireael07** [comment]
Another candidate is FixedMaterial -> whatever the 3.0 equivalent is.

**Comment by Zylann** [comment]
Also AABB -> Rect3, Matrix -> Basis, ColorFrame -> ColorRect, Mesh -> ArrayMesh

**Comment by Zylann** [comment]
Maybe also `.type == InputEvent.***` => `is ***`

**Comment by HummusSamurai** [comment]
When this is discussed for merge on IRC, I'd like to suggest that it be merged directly without waiting for all the changes to be added.

That way further additions to the conversion script can be contributed through subsequent PRs.

**Comment by akien-mga** [comment]
For opacity, you can do changes in a non-destructive way (currently you're overriding potential modulate colors) with properties:
```
self.modulate.a = var
```

**Comment by marcelofg55** [comment]
@Zylann 
> You could also rename "len" identifiers into "length", because "len" is now a reserved function keyword (if there is no other variable in the same scope named "length" already)

I thought about that one too, but it probably is a tricky one since you have to avoid getting false positives as len is a pretty short and common word. But I'll give it a try later.

> Also AABB -> Rect3, Matrix -> Basis, ColorFrame -> ColorRect, Mesh -> ArrayMesh

Are these direct conversions? I haven't used them myself in gdscript, so idk.

> Maybe also .type == InputEvent.*** => is ***

Added :).

**Comment by marcelofg55** [comment]
@akien-mga Just fixed that modulate thing, thanks for the tip :).

**Comment by reduz** [comment]
this looks great, i think, if not implemented already, animations should also be properly fixed

File: editor/editor_export_godot3.cpp [hunk 1]
```diff
@@ -1980,7 +1981,190 @@ void EditorExportGodot3::_save_config(const String &p_path) {
...
+			// Convert .set_opacity(var) => .modulate.a = var
+			regexp.compile("([ \t]*)([a-zA-Z0-9_]*)[ ]*\\.set_opacity\\((.*)\\)(.*)");
+			res = regexp.find(line);
+			if (res >= 0 && regexp.get_capture_count() == 5) {
+				line = regexp.get_capture(1) + regexp.get_capture(2) + ".modulate.a = " + regexp.get_capture(3) + regexp.get_capture(4);
+				count++;
+			}
+			regexp.clear();
+
+			// Convert var.type == InputEvent.KEY => var.get_name() == "InputEventKey"
```
Comment (Hinsbart) [inline editor/editor_export_godot3.cpp]: This is not the recommended method for type checking.
Use `var is InputEventKey` instead.

**Review by 27thLiz — CHANGES_REQUESTED** [review]

File: editor/editor_export_godot3.cpp (hunk 1 above)
Comment (marcelofg55) [reply editor/editor_export_godot3.cpp]: Just fixed it.

**Review by 27thLiz — APPROVED** [review]

File: editor/editor_export_godot3.cpp [hunk 2]
```diff
@@ -1980,7 +1981,190 @@ void EditorExportGodot3::_save_config(const String &p_path) {
...
+			regexp.compile("(.*)\\.get_opacity\\(\\)(.*)");
+			res = regexp.find(line);
+			if (res >= 0 && regexp.get_capture_count() == 3) {
+				line = regexp.get_capture(1) + ".get_modulate().a" + regexp.get_capture(2);
+				count++;
+			}
+			regexp.clear();
+
+			// Convert .set_opacity(var) => .modulate.a = var
+			regexp.compile("([ \t]*)([a-zA-Z0-9_]*)[ ]*\\.set_opacity\\((.*)\\)(.*)");
```
Comment (akien-mga) [inline editor/editor_export_godot3.cpp]: @marcelofg55 I'm doing some improvements to the script conversion, and wonder about this regexp. What's the point of `[ ]*` before `\\.set_opacity\\(`? Shouldn't it simply be `"(.*)\\.set_opacity\\((.*)\\)(.*)"`?

Comment (marcelofg55) [reply editor/editor_export_godot3.cpp]: I don't remember if there was a reason that I wrote that regexp that way, but it should work fine as you say as `(.*)\\.set_opacity\\((.*)\\)(.*)`.

**Pull request merged**